net/inet_pton.c

```diff
@@ -1,6 +1,146 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * net/inet_pton.c
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
 #include <arpa/inet.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <string.h>
 
-int inet_pton(int af, const char *restrict src, void *restrict dst)
+/* Accept only addresses in dotted-decimal format ddd.ddd.ddd.ddd */
+static int inet_pton4(const char *src, struct in_addr *dst)
 {
-	return 0;
+	in_addr_t addr4 = 0;
+	unsigned long val;
+	int i;
+	char *endptr;
+
+	for (i = 0; i < 4; i++) {
+		if (!isdigit(*src))
+			return 0;
+
+		val = strtoul(src, &endptr, 0);
+		if (endptr == src || val > 255)
+			return 0;
+		if ((*endptr != '.' && i < 3) ||
+			(*endptr != '\0' && i == 3))
+			return 0;
+
+		src = endptr + 1;
+		addr4 = addr4 << 8 | (val & 0xFFFF);
+	}
+
+	dst->s_addr = htonl(addr4);
+
+	return 1;
+}
+
+/*
+ * Accepts IPv6 addresses with optional contigous zeros abbreviation
+ * or IPv6-mapped IPv4 addresses.
+ */
+static int inet_pton6(const char *src, struct in6_addr *dst)
+{
+	struct in6_addr addr6;
+	struct in_addr addr4;
+	const int n_hextets = sizeof(addr6.s6_addr16) / sizeof(addr6.s6_addr16[0]);
+	int i, idx = 0, z = -1;
+	unsigned long val;
+	const char *endptr, *ptr = src;
+
+	memset(&addr6, 0, sizeof(addr6));
+	addr4.s_addr = 0;
+
+	while (*ptr != '\0') {
+		if (idx > n_hextets - 1)
+			return 0;
+
+		endptr = ptr;
+		while (isxdigit(*endptr))
+			endptr++;
+
+		if (ptr == endptr && *endptr == ':') {
+			/* '::' zeros abbreviation */
+			if (z != -1)
+				return 0;
+			z = idx;
+			if (ptr != src)
+				ptr = endptr + 1;
+			else if (ptr == src && *(endptr + 1) == ':')
+				ptr = endptr + 2;
+			else
+				return 0;
+		}
+		else if (*endptr == ':' || *endptr == '\0') {
+			/* Hexadecimal part */
+			val = strtoul(ptr, NULL, 16);
+			if (val > 0xFFFF)
+				return 0;
+			addr6.s6_addr16[idx++] = htons((uint16_t) val);
+
+			if (*endptr == ':')
+				ptr = endptr + 1;
+			else
+				ptr = endptr;
+		}
+		else if (*endptr == '.') {
+			/* IPv6-mapped IPv4 address */
+			if (inet_pton4(ptr, &addr4) != 1)
+				return 0;
+			memcpy(addr6.s6_addr16 + idx, &addr4.s_addr, sizeof(addr4));
+			idx += 2;
+			break;
+		}
+		else {
+			return 0;
+		}
+	}
+
+	/* Address is not full */
+	if (idx < n_hextets && z == -1)
+		return 0;
+
+	/* Address too long */
+	if (idx == n_hextets && z != -1)
+		return 0;
+
+	/* Copy address and fill the zeros string */
+	for (i = 0; i < n_hextets; i++) {
+		if (i >= z && i < n_hextets - idx + z)
+			dst->s6_addr16[i] = 0;
+		else if (i < n_hextets - idx + z)
+			dst->s6_addr16[i] = addr6.s6_addr16[i];
+		else
+			dst->s6_addr16[i] = addr6.s6_addr16[i - (n_hextets - idx)];
+	}
+
+	return 1;
+}
+
+
+int inet_pton(int af, const char *src, void *dst)
+{
+	if (src == NULL || dst == NULL)
```

Comment:
You can possibly add a check for empty `src`.
```
	if (src == NULL || dst == NULL || *src == '\0')
```

Reply:
the function would still return valid data for `src == '\0'` so it's not a strictly coding bug (as opposed to passing NULL to this function.

Regarding NULL checking - I might opt for making asserts here against NULL or allowing to crash the app by accessing the `NULL`, otherwise it might be hard to find a bug regarding invalid usage of `inet_pton`/`inet_ntop`, but that's probably broader discussion.